c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -1905,7 +2069,37 @@ void raw_delete(void* ptr) {
   caching_allocator.free(ptr);
 }
 
+std::vector<std::vector<AllocFreeEvent>> getAllocFreeEvents() {
+  return memory_tracker.get_alloc_free_events();
+}
+
+void set_mem_plan(std::vector<std::vector<AllocFreeEvent>> plan) {
+  std::vector<Block*> plan_ptrs{plan.size(), nullptr};
+  std::vector<size_t> plan_size{plan.size(), 0};
+  std::vector<unsigned int> non_rec_lengths{plan.size(), 0};
+  void* r;
+  for (unsigned int device = 0; device < plan.size(); device++) {
+    plan_size[device] = plan[device][0].size;
+    non_rec_lengths[device] = plan[device][0].ptr;
+    r = nullptr; // if ptr remains null: this means no planning for this device
+                 // because either device memory is not planned or failed to
+                 // assign enough memory for planning
+    if (plan[device].size() > 0) {
+      c10::cuda::CUDACachingAllocator::caching_allocator.malloc(
+          &r, device, plan[device][0].size, cuda::getCurrentCUDAStream(device));
+    }
+    plan_ptrs[device] =
+        c10::cuda::CUDACachingAllocator::caching_allocator.get_allocated_block(
+            r);
+  }
+  if( !memory_planner) {
+    memory_planner = new MemoryPlanner;
```

Comment:
if you want this to be a singleton, you can borrow how `RRefContext::getInstance()` was implemented.